File: minimum_rotated.py
def findMin(nums):  
        low = 0
        high = len(nums)-1
        
        if low == high:
            return nums[low]
        if nums[high]>nums[low]:
            return nums[0]
        
        while low<=high:
            mid = (low+high)//2
            if mid<high and nums[mid+1]<nums[mid]:
                return nums[mid+1]
            if mid>low and nums[mid] <nums[mid-1]:
                return nums[mid]
            if nums[high] > nums[mid]:
                high = mid-1
            else:
                low = mid+1

nums = [5, 6, 7, 8, 9, 10, 1, 2, 3]
findMin(nums)

File: Minimum_rotated.java
//Problem 153

class Solution {
    public int findMin(int[] nums) {
        int n = nums.length;
        int low = 0;
        int high = n-1;
        while(low<=high){
            int mid = low+(high-low)/2;
            if(nums[low]<nums[high]) return nums[low]; // already sorted array
            if((mid==0||nums[mid-1]>nums[mid]) && (mid==n-1||nums[mid+1]>nums[mid])){
                return nums[mid];
            }
            else if(nums[low]<=nums[mid]){ //left sorted, min lies to the right
                low = mid+1;
            }
            else{
                high = mid-1;
            }
            
        }
        return -1;
    }
}
